e2e/test/iothub/service/DirectMethodE2ETests.cs

```diff
@@ -0,0 +1,136 @@
+            errorContext.And.IsTransient.Should().BeFalse();
+        }
+
+        [TestMethod]
+        [Timeout(TestTimeoutMilliseconds)]
+        public async Task DirectMethodsClient_InvokeAsycn_ModuleDoesNotExist()
+        {
+            // arrange
+            IotHubServiceClient serviceClient = TestDevice.ServiceClient;
+
+            var methodInvocation = new DirectMethodServiceRequest("someDirectMethod");
+
+            // act
+            Func<Task> act = async () => await serviceClient.DirectMethods.InvokeAsync("someNonexistentDevice", "someNonexistentModule", methodInvocation);
+
+            // assert
+            ExceptionAssertions<IotHubServiceException> errorContext = await act.Should().ThrowAsync<IotHubServiceException>();
+            errorContext.And.StatusCode.Should().Be(System.Net.HttpStatusCode.NotFound);
+            errorContext.And.IsTransient.Should().BeFalse();
+        }
+
+        [TestMethod]
+        [Timeout(TestTimeoutMilliseconds)]
+        public async Task DirectMethodsClient_DeviceOnline_DeviceClientNotOpen()
+        {
+            // arrange
+            IotHubServiceClient serviceClient = TestDevice.ServiceClient;
+            await using TestDevice testDevice = await TestDevice.GetTestDeviceAsync(_devicePrefix);
+
+            IotHubDeviceClient deviceClient = testDevice.CreateDeviceClient();
+            await testDevice.OpenWithRetryAsync().ConfigureAwait(false);
+            var methodInvocation = new DirectMethodServiceRequest("someDirectMethod");
+
+            // ensure device client is closed
+            await deviceClient.CloseAsync();
+            
+            try
+            {
+                // act
+                Func<Task> act = async () => await serviceClient.DirectMethods.InvokeAsync(testDevice.Id, methodInvocation);
+
+                // assert
+                ExceptionAssertions<IotHubServiceException> errorContext = await act.Should().ThrowAsync<IotHubServiceException>();
+                errorContext.And.StatusCode.Should().Be(System.Net.HttpStatusCode.NotFound);
+            }
+            finally
+            {
+                await deviceClient.CloseAsync();
+            }
+        }
+
+        [TestMethod]
+        [Timeout(TestTimeoutMilliseconds)]
+        public async Task DirectMethodsClient_DeviceOnline_MethodCallbackTimesOut()
+        {
+            // arrange
+            IotHubServiceClient serviceClient = TestDevice.ServiceClient;
+            await using TestDevice testDevice = await TestDevice.GetTestDeviceAsync(_devicePrefix);
+            var methodInvocation = new DirectMethodServiceRequest("someDirectMethod")
+            {
+                ConnectionTimeout = TimeSpan.FromSeconds(3),
+                ResponseTimeout = TimeSpan.FromSeconds(3),
+            };
+            IotHubDeviceClient deviceClient = testDevice.CreateDeviceClient();
+            await deviceClient.OpenAsync();
+            await testDevice.OpenWithRetryAsync().ConfigureAwait(false);
```

Comment:
OpenWithRetryAsync calls OpenAsync with a retry policy that includes bad credentials, which is sometimes needed for brand new created devices, so you don't need to call both.

```suggestion
            await testDevice.OpenWithRetryAsync().ConfigureAwait(false);
```